### ISSUE TITLE
[cssom] Need to agree on when LinkStyle.sheet gets updated in shadow trees.

### ISSUE THREAD

**Issue by emilio** [w3c/csswg-drafts#3096]
Since it's observable via cssRules and all that stuff.

In Gecko, when a shadow host is unbound, we also disconnect the `ShadowRoot` children. This has side effects, such as losing (only temporarily) the `sheet` pointer, and losing the `disabled` state.

I was going to fix this in Gecko, when to my surprise Chrome (which is the only other implementor of `ShadowRoot.styleSheets` doesn't keep the stylesheets around in disconnected subtrees.

We should agree on when should they be kept.

Test-case (sorry, pretty messy):

```html
<!doctype html>
<div id="host"></div>
<script>
  let host = document.querySelector("#host");
  host.attachShadow({ mode: "open" }).innerHTML = `<style>* { color: red }</style><span>Should not be red</span>`;

  console.log("host.shadowRoot.styleSheets", host.shadowRoot.styleSheets);
  console.log("style.sheet", host.shadowRoot.querySelector('style').sheet);

  let sheet = host.shadowRoot.styleSheets[0];
  console.log("host.shadowRoot.styleSheets[0]", sheet);
  sheet.disabled = true;

  // assert_equals(getComputedStyle(host.shadowRoot.querySelector('span')).color, "rgb(0, 0, 0)");
  document.body.offsetTop;

  host.remove();

  console.log("host.shadowRoot.styleSheets (after removal)", host.shadowRoot.styleSheets);
  console.log("style.sheet (after removal)", host.shadowRoot.querySelector('style').sheet);
  sheet = host.shadowRoot.styleSheets[0];
  console.log("host.shadowRoot.styleSheets[0] (after removal)", sheet);
  console.log("host.shadowRoot.styleSheets[0].disabled (after removal)", sheet ? sheet.disabled : "(no sheet)");

  document.body.offsetTop;
  document.body.appendChild(host);

  document.body.offsetTop;
  let host2 = document.createElement('div');
  host2.attachShadow({ mode: "open" }).innerHTML = `<style>* { color: red }</style><span>Should not be red</span>`;
  console.log("host2.shadowRoot.styleSheets", host2.shadowRoot.styleSheets);
  console.log("style.sheet", host2.shadowRoot.querySelector('style').sheet);
  sheet = host2.shadowRoot.styleSheets[0];
  console.log("host2.shadowRoot.styleSheets[0]", sheet);
</script>
```

Or online at https://crisal.io/tmp/disabled-shadow.html.

And to be clear I'm fine speccing and implementing the Blink behavior, if we decide that accessing stylesheets in disconnected shadow roots is not worth the churn.

cc @lilles @tabatkins @rniwa

**Comment by rniwa** [comment]
Hm... Safari seems to clear the disabled-ness so that would be at least consistent. I guess the question is whether the current behaviors of Chrome/WebKit would prevent some important use cases and/or if the behavior is consistent with other scenarios not involving shadow roots. e.g. if we disconnect a subtree containing a style/link element, I presume we'd clear the stylesheet of the element?

**Comment by emilio** [comment]
Right, elements outside of the document have already a null `sheet`, so I think it's consistent.

**Comment by rniwa** [comment]
Yeah then tying this to connectedness of `style` / `link` element probably makes most sense.

**Comment by emilio** [comment]
The tricky part is that this affects `ShadowRoot.styleSheets` as well. That is, `ShadowRoot.styleSheets` for a disconnected `ShadowRoot` would always be empty. But as long as we think that's fine I'm fine with that.

**Comment by lilles** [comment]
So the HTML spec already says that sheets are not loaded/created before they are shadow-including connected, so the host2 in your test-case should certainly not be loaded per spec.

I could not find out from the spec what happens when a link/style is removed, even for non-shadow cases, but tying this to connectedness seems like the reasonable thing here.

**Comment by emilio** [comment]
The google proposals I was talking about above are `document.createStyleSheet` (https://github.com/WICG/construct-stylesheets/) and https://github.com/w3c/webcomponents/issues/468.

**Comment by hayatoito** [comment]
We have to update wpt to align the conclusion here:  https://github.com/web-platform-tests/wpt/blob/master/shadow-dom/ShadowRoot-interface.html#L92

**Comment by hayatoito** [comment]
@emilio It looks you have already updated the test. Thanks!
